x-pack/dockerlogbeat/pipelineManager/clientLogReader.go

```diff
@@ -0,0 +1,125 @@
+// you may not use this file except in compliance with the Elastic License.
+
+package pipelinemanager
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/docker/engine/api/types/plugins/logdriver"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/publisher/pipeline"
+	pb "github.com/gogo/protobuf/io"
+	"github.com/pkg/errors"
+	"github.com/tonistiigi/fifo"
+)
+
+// ClientLogger is an instance of a pipeline logger client meant for reading from a single log stream
+// There's a many-to-one relationship between clients and pipelines.
+// Each container with the same config will get its own client to the same pipeline.
+type ClientLogger struct {
+	logFile      io.ReadWriteCloser
+	client       beat.Client
+	pipelineHash string
+}
+
+// newClientFromPipeline creates a new Client logger with a FIFO reader and beat client
+func newClientFromPipeline(pipeline *pipeline.Pipeline, file, hashstring string) (*ClientLogger, error) {
+	// setup the beat client
+	settings := beat.ClientConfig{
+		WaitClose: 0,
+	}
+	settings.ACKCount = func(n int) {
+		logp.Info("Pipeline client (%s) ACKS; %v", file, n)
+	}
+	settings.PublishMode = beat.DefaultGuarantees
+	client, err := pipeline.ConnectWith(settings)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the FIFO reader client from the FIPO pipe
+	inputFile, err := fifo.OpenFifo(context.Background(), file, syscall.O_RDONLY, 0700)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error opening logger fifo: %q", file)
+	}
+	logp.Info("Created new logger for %s", file)
+
+	return &ClientLogger{logFile: inputFile, client: client, pipelineHash: hashstring}, nil
+}
+
+// Close Closes the pipeline client and reader
+func (cl *ClientLogger) Close() error {
+	logp.Info("Closing ClientLogger")
+	cl.client.Close()
+
+	return cl.logFile.Close()
+}
+
+// ConsumeAndSendLogs reads from the FIFO file and sends to the pipeline client. This will block and should be called in its own goroutine
+// TODO: Publish() can block, which is a problem. This whole thing should be two goroutines.
+func (cl *ClientLogger) ConsumeAndSendLogs() {
+	reader := pb.NewUint32DelimitedReader(cl.logFile, binary.BigEndian, 2e6)
+
+	publishWriter := make(chan logdriver.LogEntry, 500)
+
+	go cl.publishLoop(publishWriter)
+	// Clean up the reader after we're done
+	defer func() {
+
+		close(publishWriter)
+
+		err := reader.Close()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error closing FIFO reader: %s", err)
+		}
+	}()
+
+	var log logdriver.LogEntry
+	for {
+		err := reader.ReadMsg(&log)
+		if err != nil {
+			if err == io.EOF || err == os.ErrClosed || strings.Contains(err.Error(), "file already closed") {
+				cl.logFile.Close()
+				return
+			}
+			// I am...not sure why we do this
+			reader = pb.NewUint32DelimitedReader(cl.logFile, binary.BigEndian, 2e6)
```

Comment:
Where did you get this line from?

Reply:
Docker's example logging plugin does this, so every other plugin copied it and does the same thing. I'm guessing it's to 'reset' the reader in case of some undefined error, but I can't really make sense of it.

Reply:
Implementation: https://github.com/gogo/protobuf/blob/master/io/uint32.go#L114

It does not reset the reader, but returns an ErrShortBuffer. The generated reader reads the complete event into memory and then unmarshals it. The 2e6 is the max read buffer size.

Once you get a message that is potentially bigger, you are stuck. The length field has already been read (read pointer has advanced), and you MUST NOT continue reading from the reader once you get ErrShortBuffer. This potentially leaves you with another problem.

I think it is fair to rule out old big messages (keep memory usage in check). Just creating a new reader ([as done here](https://github.com/elastic/beats/pull/13761/files#diff-37bb9e7466565d4ffbe9e85058daa456R97)) without consuming any more bytes will put you in the middle of a protobuf message, giving you an invalid read. If we want a reader that can drop large messages, then we need another implementation. But the current loop will happily produce errors and reinitialize the reader over and over again after until it did find a valid length field. The way protobuf is encoded, there are lengths fields for nested structures and strings all over the place. Unmarshaling will likely fail over and over again itself.

Reply:
Ah, thanks. I'll keep reading into this.